x-pack/platform/plugins/shared/observability_ai_assistant/server/service/distributed_lock_manager/index.ts

```diff
@@ -0,0 +1,138 @@
+ * 2.0.
+ */
+
+import { Client, errors } from '@elastic/elasticsearch';
+import { Logger } from '@kbn/logging';
+
+const INDEX_NAME = '.kibana-distributed-lock-manager';
+
+export enum LockId {
+  KnowledgeBaseReindex = 'knowledge_base_reindex',
+}
+
+export interface LockDocument {
+  createdAt: string;
+  expiresAt: string;
+  meta: Record<string, any>;
+}
+
+export class LockManager {
+  constructor(private lockId: LockId, private esClient: Client, private logger: Logger) {}
+
+  /**
+   * Attempts to acquire a lock by creating a document with the given lockId.
+   * If the lock exists and is expired, it will be released and acquisition retried.
+   */
+  public async acquire(
+    options: { meta?: Record<string, any>; ttlMs?: number } = {}
+  ): Promise<boolean> {
+    const ttl = options.ttlMs ?? 3600 * 1000; // Default TTL: 1 hour.
+    const now = new Date();
+
+    try {
+      await this.esClient.index({
+        index: INDEX_NAME,
+        id: this.lockId,
+        body: {
+          createdAt: now.toISOString(),
+          expiresAt: new Date(now.getTime() + ttl).toISOString(),
+          meta: options.meta ?? {},
+        },
+        op_type: 'create', // Only create if the document does not already exist.
+      });
+      this.logger.debug(`Lock "${this.lockId}" acquired.`);
+      return true;
+    } catch (error: unknown) {
+      const isVersionConflictError =
+        error instanceof errors.ResponseError &&
+        error.body?.error?.type === 'version_conflict_engine_exception';
+
+      if (isVersionConflictError) {
+        const existingLock = await this.get();
+        if (!existingLock) {
+          // If the lock doesn't exist, it might have been purged because it expired and we will try acquiring it again.
+          return this.acquire(options);
+        }
+        return false;
+      }
+
+      throw error;
+    }
+  }
+
+  private isExpired(lock: LockDocument): boolean {
+    const { expiresAt } = lock;
+    return new Date(expiresAt) < new Date();
+  }
+
+  /**
+   * Releases a lock by deleting its document.
+   * If the document is not found, the error is ignored.
+   */
+  public async release(): Promise<void> {
+    await this.esClient.delete({ index: INDEX_NAME, id: this.lockId }, { ignore: [404] });
+    this.logger.debug(`Lock "${this.lockId}" released.`);
+  }
+
+  /**
+   * Retrieves the lock document for a given lockId.
+   * If the lock is expired, it will be released.
+   */
+  public async get(): Promise<LockDocument | undefined> {
+    const res = await this.esClient.get<LockDocument>(
+      { index: INDEX_NAME, id: this.lockId },
+      { ignore: [404] }
+    );
+
+    if (!res?._source) {
+      return;
+    }
+
+    // Check if the lock is expired.
+    if (this.isExpired(res?._source)) {
+      await this.release();
+      return;
+    }
+
+    return res._source;
+  }
+}
+
+export async function ensureTemplatesAndIndexCreated(client: Client): Promise<void> {
```

Comment:
ideally we don't create this index unless the lock is being used - you could consider looking into @kbn/storage-adapter for this

Reply:
Changed so the index is created when calling `acquire`